godel_path_planning/src/trajectory_planning.cpp

```diff
@@ -0,0 +1,188 @@
+                             double interpoint_delay,
+                             double time_offset,
+                             trajectory_msgs::JointTrajectory& trajectory)
+  {
+    typedef std::vector<descartes_core::TrajectoryPtPtr>::const_iterator JointSolutionIterator;
+    
+    // For calculating the time_from_start field of the trajectoryPoint
+    ros::Duration time_from_start(time_offset);
+
+    for (JointSolutionIterator it = solution.begin(); it != solution.end(); ++it)
+    {
+      // Retrieve actual target joint angles from the polymorphic interface function
+      std::vector<std::vector<double> > joint_angles;
+      it->get()->getJointPoses(robot_model, joint_angles);
+      const std::vector<double>& sol = joint_angles[0];
+      
+      trajectory_msgs::JointTrajectoryPoint point;
+      point.positions = sol;
+      point.velocities.resize(sol.size(), 0.0); // Fill extra fields with zeroes for now
+      point.accelerations.resize(sol.size(), 0.0);
+      point.effort.resize(sol.size(), 0.0);
+      point.time_from_start = time_from_start;
+
+      // add trajectory point to array
+      trajectory.points.push_back(point);
+
+      // increment time so far by next duration
+      time_from_start += ros::Duration(interpoint_delay);
+    }
+
+    return true;
+  }
+
+
+} // end of anon namespace
+
+bool godel_path_planning::generateTrajectory(const godel_msgs::TrajectoryPlanning::Request& req,
+                                             trajectory_msgs::JointTrajectory& trajectory)
+{
+  using descartes_core::TrajectoryPtPtr;
+
+  if (req.path.points.empty())
+  {
+    ROS_WARN("%s: recieved trajectory planning request with no points.", __FUNCTION__);
+    return false;
+  }
+
+  // Note that there is both a descartes_core::RobotModel and a moveit RobotModel
+  robot_model_loader::RobotModelLoader robot_model_loader("robot_description");
+  robot_state::RobotStatePtr kinematic_state (new robot_state::RobotState(robot_model_loader.getModel()));
+
+  descartes_core::RobotModelConstPtr robot_model = createRobotModel(kinematic_state, //group.getCurrentState(),
+                                                                    req.group_name, req.tool_frame, 
+                                                                    req.world_frame, req.iterations);
+
+  std::vector<TrajectoryPtPtr> graph_points;
+  graph_points.reserve(req.path.points.size());
+
+  // Create cartesian trajectory points out of the tool processing path
+  for (size_t i = 0; i < req.path.points.size(); ++i)
+  {
+    // compute the absolute transform of a given point given its 
+    // reference plane and relative position to that plane
+    tf::Transform point_tf = createNominalTransform(req.path.reference, req.path.points[i]);
+    graph_points.push_back(tfToAxialTrajectoryPt(point_tf, req.angle_discretization));
+  }
+
+  descartes_core::PathPlannerBasePtr planner (new descartes_planner::DensePlanner);
+  planner->initialize(robot_model);
+
```

Comment:
I agree that reloading the robot_description for every request is unnecessary (as I think we can assume it does not change) and probably wastes a fair bit of time. I disagree as far as the planner and robot_state goes. I can't imagine that the cost of initializing these objects comes within two orders of magnitude of the cost of building/searching the graph. 

With that in mind I'd prefer to keep the current architecture that only uses free functions. The flow of data and control is very clear, you don't have to consult a seperate header file, there's no lingering state to be managed between calls, the function is thread-safe, and it can be dropped into other pieces of code without having to write any glue interface.

I'm probably being a bit ardent with this one, but I believe functional-programming esque code has huge benefits when applied to problems and code-bases as big as ours.